File: src/net/dinkla/raytracer/math/BBox.java
package net.dinkla.raytracer.math;

import net.dinkla.raytracer.objects.GeometricObject;

import java.util.List;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 10.04.2010
 * Time: 15:28:49
 * To change this template use File | Settings | File Templates.
 */
public class BBox {

    public final Point3DF p;
    public final Point3DF q;

    public BBox() {
        p = null;
        q = null;
    }
    
    public BBox(final Point3DF p, final Point3DF q) {
        if (null != p && null != q) {
            if (p.x > q.x || p.y > q.y || p.z > q.z) {
                int a = 2;
            }
            assert p.x <= q.x && p.y <= q.y && p.z <= q.z;
        }
        this.p = p;
        this.q = q;
    }

    /**
     * Is Point r inside the bounding box?
     *
     * @param r     A point.
     * @return      True, if the point r is inside the bounding box.
     */
    public boolean inside(final Point3DF r) {
        boolean isX = r.x > p.x && r.x < q.x;
        boolean isY = r.y > p.y && r.y < q.y;
        boolean isZ = r.z > p.z && r.z < q.z;
        return (isX && isY && isZ);
    }

    static public class Hit {
        public final float t0;
        public final float t1;
        public final boolean isHit;

        public Hit(final float t0, final float t1) {
            this.t0 = t0;
            this.t1 = t1;
            isHit = t0 < t1 && t1 > MathUtils.K_EPSILON;
        }

        public Hit() {
            isHit = false;
            t0 = Float.NaN;
            t1 = Float.NaN;
        }
    }

    public Hit hitX(final Ray ray) {
        if (null == p && null == q) {
            return new Hit();
        }
        float tx_min, ty_min, tz_min;
        float tx_max, ty_max, tz_max;

        float a = 1.0f / ray.d.x;
        if (a >= 0) {
            tx_min = (p.x - ray.o.x) * a;
            tx_max = (q.x - ray.o.x) * a;
        } else {
            tx_min = (q.x - ray.o.x) * a;
            tx_max = (p.x - ray.o.x) * a;
        }

        float b = 1.0f / ray.d.y;
        if (b >= 0) {
            ty_min = (p.y - ray.o.y) * b;
            ty_max = (q.y - ray.o.y) * b;
        } else {
            ty_min = (q.y - ray.o.y) * b;
            ty_max = (p.y - ray.o.y) * b;
        }

        float c = 1.0f / ray.d.z;
        if (c >= 0) {
            tz_min = (p.z - ray.o.z) * c;
            tz_max = (q.z - ray.o.z) * c;
        } else {
            tz_min = (q.z - ray.o.z) * c;
            tz_max = (p.z - ray.o.z) * c;
        }

        float t0, t1;

        // find largest entering t value
        if (tx_min > ty_min) {
            t0 = tx_min;
        } else {
            t0 = ty_min;
        }

        if (tz_min > t0) {
            t0 = tz_min;
        }

        // find smallest exiting t value
        if (tx_max < ty_max) {
            t1 = tx_max;
        } else {
            t1 = ty_max;
        }
        if (tz_max < t1) {
            t1 = tz_max;
        }

        return new Hit(t0, t1);
    }


    public boolean hit(final Ray ray) {
        if (null == p && null == q) {
            return false;
        }

        float tx_min, ty_min, tz_min;
        float tx_max, ty_max, tz_max;

        float a = 1.0f / ray.d.x;
        if (a >= 0) {
            tx_min = (p.x - ray.o.x) * a;
            tx_max = (q.x - ray.o.x) * a;
        } else {
            tx_min = (q.x - ray.o.x) * a;
            tx_max = (p.x - ray.o.x) * a;
        }

        float b = 1.0f / ray.d.y;
        if (b >= 0) {
            ty_min = (p.y - ray.o.y) * b;
            ty_max = (q.y - ray.o.y) * b;
        } else {
            ty_min = (q.y - ray.o.y) * b;
            ty_max = (p.y - ray.o.y) * b;
        }

        float c = 1.0f / ray.d.z;
        if (c >= 0) {
            tz_min = (p.z - ray.o.z) * c;
            tz_max = (q.z - ray.o.z) * c;
        } else {
            tz_min = (q.z - ray.o.z) * c;
            tz_max = (p.z - ray.o.z) * c;
        }

        float t0, t1;

        // find largest entering t value
        if (tx_min > ty_min) {
            t0 = tx_min;
        } else {
            t0 = ty_min;
        }

        if (tz_min > t0) {
            t0 = tz_min;
        }

        // find smallest exiting t value
        if (tx_max < ty_max) {
            t1 = tx_max;
        } else {
            t1 = ty_max;
        }
        if (tz_max < t1) {
            t1 = tz_max;
        }

        return (t0 < t1 && t1 > MathUtils.K_EPSILON);
    }

    static public BBox create(final Point3DF v0, final Point3DF v1, final Point3DF v2) {
        float x0 = Float.POSITIVE_INFINITY;
        float x1 = Float.NEGATIVE_INFINITY;
        if (v0.x < x0) {
            x0 = v0.x;
        }
        if (v1.x < x0) {
            x0 = v1.x;
        }
        if (v2.x < x0) {
            x0 = v2.x;
        }
        if (v0.x > x1) {
            x1 = v0.x;
        }
        if (v1.x > x1) {
            x1 = v1.x;
        }
        if (v2.x > x1) {
            x1 = v2.x;
        }
        float y0 = Float.POSITIVE_INFINITY;
        float y1 = Float.NEGATIVE_INFINITY;
        if (v0.y < y0) {
            y0 = v0.y;
        }
        if (v1.y < y0) {
            y0 = v1.y;
        }
        if (v2.y < y0) {
            y0 = v2.y;
        }
        if (v0.y > y1) {
            y1 = v0.y;
        }
        if (v1.y > y1) {
            y1 = v1.y;
        }
        if (v2.y > y1) {
            y1 = v2.y;
        }
        float z0 = Float.POSITIVE_INFINITY;
        float z1 = Float.NEGATIVE_INFINITY;
        if (v0.z < z0) {
            z0 = v0.z;
        }
        if (v1.z < z0) {
            z0 = v1.z;
        }
        if (v2.z < z0) {
            z0 = v2.z;
        }
        if (v0.z > z1) {
            z1 = v0.z;
        }
        if (v1.z > z1) {
            z1 = v1.z;
        }
        if (v2.z > z1) {
            z1 = v2.z;
        }
        return new BBox(new Point3DF(x0 - MathUtils.K_EPSILON, y0 - MathUtils.K_EPSILON, z0 - MathUtils.K_EPSILON),
                new Point3DF(x1 + MathUtils.K_EPSILON, y1 + MathUtils.K_EPSILON, z1 + MathUtils.K_EPSILON));
    }

    public static BBox create(final List<GeometricObject> objects) {
        if (objects.size() > 0) {
            Point3DF p = PointUtilities.minCoordinates(objects);
            Point3DF q = PointUtilities.maxCoordinates(objects);
            return new BBox(p, q);
        } else {
            return new BBox();
        }
    }

    public float getVolume() {
        if (null == p) {
            return 0;
        } else {
            Vector3DF width = q.minus(p);
            return width.x * width.y * width.z;
        }
    }

    public boolean isContainedIn(final BBox bbox) {
        boolean bX = bbox.p.x <= p.x && q.x <= bbox.q.x;
        boolean bY = bbox.p.y <= p.y && q.y <= bbox.q.y;
        boolean bZ = bbox.p.z <= p.z && q.z <= bbox.q.z;
        return bX && bY && bZ;
    }    

    /**
     * Restrict to bbox.
     * @param bbox
     * @return
     */
    public BBox clipTo(final BBox bbox) {
        if (isContainedIn(bbox)) {
            return this;
        }
        float px = Math.max(p.x, bbox.p.x);
        float py = Math.max(p.y, bbox.p.y);
        float pz = Math.max(p.z, bbox.p.z);

        float qx = Math.min(q.x, bbox.q.x);
        float qy = Math.min(q.y, bbox.q.y);
        float qz = Math.min(q.z, bbox.q.z);

        return new BBox(new Point3DF(px, py, pz), new Point3DF(qx, qy, qz));
    }


    public BBox splitLeft(final Axis axis, final float split) {
        switch(axis) {
            case X:
                return new BBox(p, new Point3DF(split, q.y, q.z));
            case Y:
                return new BBox(p, new Point3DF(q.x, split, q.z));
            case Z:
                return new BBox(p, new Point3DF(q.x, q.y, split));
        }
        return null;
    }

    public BBox splitRight(final Axis axis, final float split) {
        switch(axis) {
            case X:
                return new BBox(new Point3DF(split, p.y, p.z), q);
            case Y:
                return new BBox(new Point3DF(p.x, split, p.z), q);
            case Z:
                return new BBox(new Point3DF(p.x, p.y, split), q);
        }
        return null;
    }

    @Override
    public boolean equals(Object obj) {
        if (null != obj) {
            if (obj instanceof BBox) {
                BBox o = (BBox) obj;
                return p.equals(o.p) && q.equals(o.q);
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return "BBox " + p.toString() + "-" + q.toString();
    }
    
}

File: src/main/kotlin/net/dinkla/raytracer/math/BBox.kt
package net.dinkla.raytracer.math

import net.dinkla.raytracer.objects.GeometricObject
import java.util.*
import kotlin.collections.ArrayList

class BBox(val p: Point3D = Point3D.ORIGIN, val q: Point3D = Point3D.ORIGIN) {

    init {
        assert(p.x <= q.x && p.y <= q.y && p.z <= q.z)
    }

    val volume: Double
        get() = (q - p).volume()

    fun inside(r: Point3D): Boolean {
        val isX = r.x > p.x && r.x < q.x
        val isY = r.y > p.y && r.y < q.y
        val isZ = r.z > p.z && r.z < q.z
        return isX && isY && isZ
    }

    class Hit {
        val t0: Double
        val t1: Double
        val isHit: Boolean

        constructor(t0: Double, t1: Double) {
            this.t0 = t0
            this.t1 = t1
            isHit = t0 < t1 && t1 > MathUtils.K_EPSILON
        }

        constructor() {
            isHit = false
            t0 = java.lang.Double.NaN
            t1 = java.lang.Double.NaN
        }
    }

    fun hitX(ray: Ray): Hit {
        val txMin: Double
        val tyMin: Double
        val tzMin: Double
        val txMax: Double
        val tyMax: Double
        val tzMax: Double

        val a = 1.0 / ray.direction.x
        if (a >= 0) {
            txMin = (p.x - ray.origin.x) * a
            txMax = (q.x - ray.origin.x) * a
        } else {
            txMin = (q.x - ray.origin.x) * a
            txMax = (p.x - ray.origin.x) * a
        }

        val b = 1.0 / ray.direction.y
        if (b >= 0) {
            tyMin = (p.y - ray.origin.y) * b
            tyMax = (q.y - ray.origin.y) * b
        } else {
            tyMin = (q.y - ray.origin.y) * b
            tyMax = (p.y - ray.origin.y) * b
        }

        val c = 1.0 / ray.direction.z
        if (c >= 0) {
            tzMin = (p.z - ray.origin.z) * c
            tzMax = (q.z - ray.origin.z) * c
        } else {
            tzMin = (q.z - ray.origin.z) * c
            tzMax = (p.z - ray.origin.z) * c
        }

        var t0: Double
        var t1: Double

        // find largest entering t value
        t0 = if (txMin > tyMin) {
            txMin
        } else {
            tyMin
        }

        if (tzMin > t0) {
            t0 = tzMin
        }

        // find smallest exiting t value
        t1 = if (txMax < tyMax) {
            txMax
        } else {
            tyMax
        }
        if (tzMax < t1) {
            t1 = tzMax
        }

        return Hit(t0, t1)
    }


    fun hit(ray: Ray): Boolean {
        val txMin: Double
        val tyMin: Double
        val tzMin: Double
        val txMax: Double
        val tyMax: Double
        val tzMax: Double

        val a = 1.0 / ray.direction.x
        if (a >= 0) {
            txMin = (p.x - ray.origin.x) * a
            txMax = (q.x - ray.origin.x) * a
        } else {
            txMin = (q.x - ray.origin.x) * a
            txMax = (p.x - ray.origin.x) * a
        }

        val b = 1.0 / ray.direction.y
        if (b >= 0) {
            tyMin = (p.y - ray.origin.y) * b
            tyMax = (q.y - ray.origin.y) * b
        } else {
            tyMin = (q.y - ray.origin.y) * b
            tyMax = (p.y - ray.origin.y) * b
        }

        val c = 1.0 / ray.direction.z
        if (c >= 0) {
            tzMin = (p.z - ray.origin.z) * c
            tzMax = (q.z - ray.origin.z) * c
        } else {
            tzMin = (q.z - ray.origin.z) * c
            tzMax = (p.z - ray.origin.z) * c
        }

        var t0: Double
        var t1: Double

        // find largest entering t value
        t0 = if (txMin > tyMin) {
            txMin
        } else {
            tyMin
        }

        if (tzMin > t0) {
            t0 = tzMin
        }

        // find smallest exiting t value
        t1 = if (txMax < tyMax) {
            txMax
        } else {
            tyMax
        }
        if (tzMax < t1) {
            t1 = tzMax
        }

        return t0 < t1 && t1 > MathUtils.K_EPSILON
    }

    private fun isContainedIn(bbox: BBox): Boolean {
        return bbox.p.x <= p.x && q.x <= bbox.q.x
                && bbox.p.y <= p.y && q.y <= bbox.q.y
                && bbox.p.z <= p.z && q.z <= bbox.q.z
    }

    fun clipTo(bbox: BBox): BBox {
        if (isContainedIn(bbox)) {
            return this
        }
        val px = Math.max(p.x, bbox.p.x)
        val py = Math.max(p.y, bbox.p.y)
        val pz = Math.max(p.z, bbox.p.z)

        val qx = Math.min(q.x, bbox.q.x)
        val qy = Math.min(q.y, bbox.q.y)
        val qz = Math.min(q.z, bbox.q.z)

        return BBox(Point3D(px, py, pz), Point3D(qx, qy, qz))
    }

    fun splitLeft(axis: Axis, split: Double): BBox? {
        return when (axis) {
            Axis.X -> BBox(p, Point3D(split, q.y, q.z))
            Axis.Y -> BBox(p, Point3D(q.x, split, q.z))
            Axis.Z -> BBox(p, Point3D(q.x, q.y, split))
        }
    }

    fun splitRight(axis: Axis, split: Double): BBox? {
        return when (axis) {
            Axis.X -> BBox(Point3D(split, p.y, p.z), q)
            Axis.Y -> BBox(Point3D(p.x, split, p.z), q)
            Axis.Z -> BBox(Point3D(p.x, p.y, split), q)
        }
    }

    override fun equals(other: Any?): Boolean {
        return if (other == null || other !is BBox) {
            false
        } else {
            val bbox: BBox = other as BBox
            p == bbox.p && q == bbox.q
        }
    }

    override fun hashCode(): Int = Objects.hash(p, q)

    override fun toString(): String = "BBox($p, $q)"

    companion object {

        fun create(v0: Point3D, v1: Point3D, v2: Point3D): BBox {
            var x0 = java.lang.Double.POSITIVE_INFINITY
            var x1 = java.lang.Double.NEGATIVE_INFINITY
            if (v0.x < x0) {
                x0 = v0.x
            }
            if (v1.x < x0) {
                x0 = v1.x
            }
            if (v2.x < x0) {
                x0 = v2.x
            }
            if (v0.x > x1) {
                x1 = v0.x
            }
            if (v1.x > x1) {
                x1 = v1.x
            }
            if (v2.x > x1) {
                x1 = v2.x
            }
            var y0 = java.lang.Double.POSITIVE_INFINITY
            var y1 = java.lang.Double.NEGATIVE_INFINITY
            if (v0.y < y0) {
                y0 = v0.y
            }
            if (v1.y < y0) {
                y0 = v1.y
            }
            if (v2.y < y0) {
                y0 = v2.y
            }
            if (v0.y > y1) {
                y1 = v0.y
            }
            if (v1.y > y1) {
                y1 = v1.y
            }
            if (v2.y > y1) {
                y1 = v2.y
            }
            var z0 = java.lang.Double.POSITIVE_INFINITY
            var z1 = java.lang.Double.NEGATIVE_INFINITY
            if (v0.z < z0) {
                z0 = v0.z
            }
            if (v1.z < z0) {
                z0 = v1.z
            }
            if (v2.z < z0) {
                z0 = v2.z
            }
            if (v0.z > z1) {
                z1 = v0.z
            }
            if (v1.z > z1) {
                z1 = v1.z
            }
            if (v2.z > z1) {
                z1 = v2.z
            }
            return BBox(Point3D(x0 - MathUtils.K_EPSILON, y0 - MathUtils.K_EPSILON, z0 - MathUtils.K_EPSILON),
                    Point3D(x1 + MathUtils.K_EPSILON, y1 + MathUtils.K_EPSILON, z1 + MathUtils.K_EPSILON))
        }

        fun create(objects: ArrayList<GeometricObject>): BBox {
            if (objects.size > 0) {
                val (p0, p1) = PointUtilities.minMaxCoordinates(objects)
                return BBox(p0, p1)
            } else {
                return BBox()
            }
        }
    }

}
